mitodl/micromasters
# Upgrade to Elasticsearch 5

#### What are the relevant tickets?
Fixes #2867 
Fixes #3794 

#### What's this PR do?
- upgrades docker container to Elasticsearch 5, and turns on an option to prevent specifying an index in the payload (it would go in the URL instead, which is how we currently do it)
 - updates elasticsearch python libraries and changes code accordingly
 - adds support for one index per type instead of allowing multiple doc types per index. In Elasticsearch 6 this will be enforced so we need to make this change at some point.
 - The percolate doc type is also removed and we need to switch to using a percolate field instead.
 - Adds a fix for percolate queries which use an unsupported syntax in Elasticsearch 5. We'll probably want to remove this fix after we reindex on production along with the rest of the legacy code.

#### How should this be manually tested?
 - First, check out master and make sure the search works as expected. 
 - Stop your docker containers. Check out this branch and run `docker-compose up`. Note: do not run recreate_index yet. This is testing backwards compatibility.
 - In a django shell run:

        from search.connection import get_conn
        conn = get_conn(verify=False)
        conn.info()

 - Verify that the Elasticsearch version is 5.5.2.
 - Go to `/learners` and verify that the search works fine
 - Update your profile. Verify on the django logs and the browser logs that no exceptions were thrown.

Now run `./manage.py recreate_index`:

 -  If you can try refreshing the learner search page while this is running. You should not notice any change in behavior.
 - Update your profile again and verify in the browser and django logs that everything looks fine

#### Any background context you want to provide?

Deployment:
 - Deploy this PR. It will work with the old and new schemas, and the old and new versions of Elasticsearch
 - Then, upgrade Elasticsearch to 5.x.
 - Finally, run ./manage.py recreate_index
